### PR TITLE
[profiler][cupti] Record graph node dependencies and draw them in the JSON export

### PR DESCRIPTION
**Pull request by dolpm**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #187898
* #188939
* #190851
* __->__ #190850

Introduce the CUDA-graph node->node dependency infrastructure and render it in the
default chrome/JSON export. A generic post-instantiate handler registry lives in
torch.cuda.graphs (register_graph_instantiate_hook; fired at the end of
CUDAGraph.instantiate()), so the graph code carries no consumer knowledge.

Recording is owned by a process-global recorder (torch.profiler._cupti._graph_deps),
armed once from the profiler when enable_graph_dependencies is set -- at profiler
construction, before the training loop captures its CUDA graphs. Its instantiate
handler reads each exec graph's node dependency edges -- keyed by tools_id (== the
CUPTI graph_node_id, joining to profiler kernel records) -- from get_graph_data()
into a persistent map. Arming early and persisting the map is required: a graph is
instantiated once (during warm-up) and thereafter only replayed, while the per-window
observer is created at prepare_trace (after warm-up) and torn down between windows, so
its own hook would never observe an instantiate(). keep_graph is irrelevant -- the
handler runs during instantiate() while the template is still live. Each observer's
dependency resolver reads the shared map by reference. The edges are threaded into the
trace window as graph_deps; _trace_window_entries then draws one flow (predecessor end
-> successor start) per edge, resolved within the same replay (correlation_id) so
arrows never cross replays, on the ops' display lanes.

A single instantiate() can drive several get_graph_data() consumers (the dependency
handler above plus an export_graph_data() dump hook), so get_graph_data() is cached
for the duration of the post-instantiate hooks and dropped immediately afterwards:
the topology is queried once per instantiate and never retained for the graph's
lifetime.

This is the format-agnostic base: the .pftrace export inherits graph_deps and only
changes the encoding (native GpuRenderStageEvent.event_wait_ids) in a later commit.

Test Plan:
    python test/profiler/test_cupti_monitor.py \
        TestCuptiRecords.test_graph_dependency_flows_json

    python test/test_cuda_graph_utils.py TestGetGraphData

Full cupti-monitor suite green at this commit:

    python test/profiler/test_cupti_monitor.py TestCuptiRecords

Authored with the assistance of an AI coding assistant.

### CODE REVIEW

File: torch/profiler/_cupti/observers/base.py [hunk 1]
```diff
@@ -151,6 +184,52 @@ def __init__(
...
+        graph_node_id, joining to profiler kernel records). The handler captures only the map
+        (never self), so the global registry cannot pin this observer; the handle is removed in
+        close()."""
+        from torch.cuda.graphs import register_graph_instantiate_hook
+
+        deps = self._graph_dependencies
+
+        def handler(torch_cuda_graph: Any) -> None:
+            try:
+                nodes = torch_cuda_graph.get_graph_data()["nodes"]
```
Comment (ngimel) [inline torch/profiler/_cupti/observers/base.py]: heh maybe we need to cache graph data, in case we dump it and use it here

File: torch/profiler/_cupti/observers/base.py [hunk 2]
```diff
@@ -151,6 +184,52 @@ def __init__(
...
+            except (RuntimeError, AttributeError, KeyError):
+                return
+            tid_by_index = [n["tools_id"] for n in nodes]
+            deps.update(
+                {
+                    n["tools_id"]: [tid_by_index[i] for i in n["dependencies"]]
+                    for n in nodes
+                    if n["dependencies"]
+                }
```
Comment (ngimel) [inline torch/profiler/_cupti/observers/base.py]:
```suggestion
            deps.update(
                {
                    n["tools_id"]: [nodes[i]["tools_id"] for i in n["dependencies"]]
                    for n in nodes
                    if n["dependencies"]
                }
```
nodes is a list already, no point in making another tid_by_index list, and tid_by_index is a confusing name anyway (implies it's a dict, even though `by_index` is more like `ordered`

**Review by ngimel — APPROVED** [review]
This is super cool

File: torch/profiler/_cupti/observers/base.py (hunk 1 above)
Comment (dolpm) [reply torch/profiler/_cupti/observers/base.py]: it's a tradeoff between running it 2x and having to keep it alive for the duration of the graph in the cache. idk which is better lol

Comment (ngimel) [reply torch/profiler/_cupti/observers/base.py]: We can destroy post instantiation and hooks
